pkg/gofr/datasource/file/ftp/file.go

```diff
@@ -0,0 +1,324 @@
+}
+
+// Scan decodes the next JSON object into the provided structure.
+func (j jsonReader) Scan(i interface{}) error {
+	return j.decoder.Decode(&i)
+}
+
+// Next checks if there is another line available in the text file.
+func (f textReader) Next() bool {
+	return f.scanner.Scan()
+}
+
+// Scan scans the next line from the text file into the provided pointer to string.
+func (f textReader) Scan(i interface{}) error {
+	switch target := i.(type) {
+	case *string:
+		*target = f.scanner.Text()
+		return nil
+	default:
+		return errNotPointer
+	}
+}
+
+// Close closes the FTP file connection.
+func (f *ftpFile) Close() error {
+	// Close the FTP response.
+	err := f.response.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Name returns the name of the file.
+func (f *ftpFile) Name() string {
+	return f.name
+}
+
+// Read reads data from the FTP file into the provided byte slice.
+func (f *ftpFile) Read(p []byte) (n int, err error) {
+	// Retrieve the file content from FTP server and read into p
+	r, err := f.conn.Retr(f.path)
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	for i := range b {
+		p[i] = b[i]
+	}
+
+	f.offset = int64(len(b))
+
+	return len(b), nil
+}
+
+// ReadAt reads data from the FTP file starting at the specified offset.
+func (f *ftpFile) ReadAt(p []byte, off int64) (n int, err error) {
+	// Retrieve file content from the specified offset
+	resp, err := f.conn.RetrFrom(f.path, uint64(off))
+	if err != nil {
+		return 0, err
+	}
+
+	b, err := io.ReadAll(resp)
+	if err != nil {
+		return 0, err
+	}
+
+	// Deep Copy
+	for i := range b {
+		p[i] = b[i]
+	}
+
+	// setting current offset of the file
+	f.offset = off + int64(len(b))
+
+	return len(b), nil
+}
+
+// Seek sets the offset for the next Read or ReadAt operation.
+func (f *ftpFile) Seek(offset int64, whence int) (int64, error) {
+	var (
+		err error
+		r   ftpResponse
+	)
+
+	// Handle Seek operation based on whence parameter
+	r, err = f.conn.Retr(f.path)
+	if err != nil {
+		return 0, err
+	}
+
+	p, err := io.ReadAll(r)
+	if err != nil {
+		return 0, err
+	}
+
+	switch whence {
+	case io.SeekStart:
+		// Seek from the beginning of the file
+		if offset < 0 || offset > int64(len(p)) {
+			return 0, datasource.ErrOutOfRange
+		}
+
+		return offset, nil
+
+	case io.SeekEnd:
+		// Seek from the end of the file
+		if offset > 0 || offset < int64(-len(p)) {
+			return 0, datasource.ErrOutOfRange
+		}
+
+		return int64(len(p)) + offset, nil
+
+	case io.SeekCurrent:
+		// Seek from the current offset
+		if f.offset+offset >= int64(len(p)) || f.offset+offset < 0 {
+			return 0, datasource.ErrOutOfRange
+		}
+
+		return f.offset + offset, nil
+
+	default:
+	}
+
+	return 0, datasource.ErrOutOfRange
+}
+
+// Write writes data to the FTP file.
+func (f *ftpFile) Write(p []byte) (n int, err error) {
+	// Delete the existing file
+	err = f.conn.Delete(f.path)
+	if err != nil {
+		return 0, err
+	}
```

Comment:
😮 This is a very unexpected action to be taken!

Note that the semantics of “delete and replace” is very much different from “truncate and rewrite”.

Not to mention the race condition of some other client or task recreating this file between this delete and the `Stor` below.

Also, what if the `f.path` does not already exist? Would this error out with a message that references deleting the file? Because that would be a very surprising error message to receive as the caller.

Reply:
I guess I will use `StorFrom` instead.